tests/test_core.py

```diff
@@ -1639,6 +1639,20 @@ class Polymorphic {virtual void member(){}};
 }
 ''', 'exception caught: std::bad_typeid')
 
+  @with_both_eh_sjlj
+  def test_terminate_abort(self):
+    # std::terminate eventually calls abort(). We used ti implement abort() with
+    # throwing a JS exception, but this can be again caught by std::terminate's
+    # cleanup and cause an infinite loop. When Wasm EH is enabled, abort() is
+    # implemented by a trap now.
+    err = self.do_run(r'''
+#include <exception>
+int main() {
+  std::terminate();
```

Comment:
Can you just call C abort() here to be more direct?

Reply:
I think this needs a little more context. 
`noexcept` function shouldn't throw, so `noexcept` function code generation is to `invoke` every function call in those functions and in case they throw, call `std::terminate`. This codegen comes from clang and native platform does this too. So in wasm, they become something like
```wasm
try
  function body
catch_all
  call std::terminate
end
```

`std::terminate` calls `std::__terminate`. Both of `std::terminate` and `std::__terminate` are `noexcept`. So that means their code is structured like that, which sounds like self-calling, but normally no function calls in those functions should ever throw, so that's fine. But in our case, `abort` ends up throwing, which is a problem.

The function body of `__terminate` eventually calls JS `abort`, and ends up here: https://github.com/emscripten-core/emscripten/blob/970998b2670a9bcf39d31e2b01db571089955add/src/preamble.js#L605-L623

This ends up throwing a JS exception. This is basically just a foreign exception from the wasm perspective, and is caught by `catch_all`, and calls `std::terminate` again. And the whole process continues until the call stack is exhausted.

What #9730 tried to do was throwing a trap, because Wasm `catch`/`catch_all` don't catch traps. Traps become `RuntimeError`s after they hit a JS frame. To be consistent, we decided `catch`/`catch_all` shouldn't catch them after they become `RuntimeError`s. That's the reason #9730 changed the code to throw not just any random thing but `RuntimeError`. But somehow we decided that we make that trap distinction not based on `RuntimeError` class but some hidden field (https://github.com/WebAssembly/exception-handling/issues/89#issuecomment-559533953).

So to answer your original question, if we use `abort()`, it will just throw `RuntimeError` and end there without a problem, because there's no `noexcept` function involved and there will be no `catch_all` and `call std::terminate`. We need `std::terminate` to show the call stack exhausting behavior.

Reply:
I think I understand..  Perhaps a more direct test would be that its not possible to catch exceptions thrown by `abort()`.  e.g.:

```
try {
  abort();
} catch (...) {
  print('should never see this\n');
}
```

Here we are testing that whatever the low level abort does is not catchable.. which I think is important part of this change.

Reply:
This is what I wrote and deleted. Actually I I was confused and this is not caught by the `catch`, because our `catch (...)` only catches C++ exceptions for now. I think whether `catch (...)` should catch foreign exceptions or not is something we should decide in the tool convention, and current impl is not catching it. So JS exceptions are only caught by the cleanups, which run destructors.

So I think removing `noexcept` from `std::terminate` and `std::__terminate` work; it will unwind the stack and run the destructors, but that's maybe OK with users..? (This is one of the things we discussed in the meeting this morning; whether `abort` should unwind or not)